entities/src/main/java/org/odk/collect/entities/LocalEntityUseCases.kt

```diff
@@ -70,11 +77,33 @@ object LocalEntityUseCases {
         val newAndUpdated = ArrayList<Entity>()
         csvParser.use {
             it.forEach { record ->
-                val serverEntity = parseEntityFromRecord(record, list) ?: return
+                val serverEntity = parseEntityFromRecord(record) ?: return
                 val existing = missingFromServer.remove(serverEntity.id)
 
-                if (existing == null || existing.version <= serverEntity.version) {
-                    newAndUpdated.add(serverEntity.copy(branchId = UUID.randomUUID().toString()))
+                if (existing == null) {
+                    newAndUpdated.add(
+                        Entity.New(
+                            serverEntity.id,
+                            serverEntity.label,
+                            serverEntity.version,
+                            serverEntity.properties.toList(),
+                            state = Entity.State.ONLINE,
+                            trunkVersion = serverEntity.version,
+                            branchId = UUID.randomUUID().toString()
+                        )
+                    )
+                } else if (existing.version <= serverEntity.version) {
+                    if (existing.isDirty() || serverEntity.version > existing.version) {
```

Comment:
I'm not sure if the logic here is correct...
If `existing.version` == `serverEntity.version` in line 95
and her `existing.version` > `existing.trunkVersion` we should update the entity with server entity?

Reply:
It's correct I'm pretty sure, but it is confusingly written (flipping the `existing`/`serverEntity` sides for example). I'll rewrite it.

Reply:
It seems I might be missing some understanding about entities. Could you explain how this works or even better point me to any relevant documentation or issue that describes this behavior?

To ensure we're on the same page, here's the scenario:
1. A form with entities is downloaded from the server.
2. One of the entities is updated locally.
3. The form with entities is redownloaded (same version as in step 1 nothing is changed).
4. The updated entity is missing, and the list of entities appears as it did in step 1.

The scenario seems unusual to me, but there might be some rationale behind it...

Reply:
> The updated entity is missing, and the list of entities appears as it did in step 1.

No in this scenario, the local update would remain in place as the local entity version would be higher than the server entity version. We only override the local with the server's data if the server's versions is greater than or equal to.

The line commented on here is in branch where there is an existing local entity with the same ID as a server entity, and the server entity's version is equal to or greater than the local one. In both these cases we want to override with the server's data (even if this clobbers a local update) - this works better with the `trunkVersion`/`branchId` concept than preferring the local version. The `isDirty` check stops us from saving things to the DB that are already there - we don't need to both saving if we have the same version, and we haven't made an update.

Reply:
Ahh right now I understand thanks!